### PR TITLE
Optimize

### PR DESCRIPTION
**Pull request by Cormac-C**
Got rid of mixed precision
Adjusted the reward function to handle tensors properly
Added use of `padding_mask` in the objective function and multiplying by log_probs

### CODE REVIEW

File: grpo.py [hunk 1]
```diff
@@ -299,25 +280,28 @@ def compute_log_probs(
...
     ), "Input IDs must have the same batch size as query IDs and generated IDs"
 
     input_ids = input_ids.to(policy.device)
     attention_mask = input_ids.ne(tokenizer.pad_token_id).long().to(policy.device)
+    logger.info(f"Attention mask shape: {attention_mask.shape}")
+    logger.info(f"Amount of tokens: {attention_mask.sum()}")
 
     # Run forward pass to get the logits
     outputs = policy(input_ids=input_ids, attention_mask=attention_mask)
     logits = outputs.logits
```
Comment (abhijeetpraveen) [inline grpo.py:299]: In the aha example, they also scale the logits with temperature before the softmax [here](https://github.com/McGill-NLP/nano-aha-moment/blob/0ce62ece2681eefad8041b9336fc7d2cd1a3687a/utils.py#L157). 
They explain it in their [video](https://youtu.be/dxhCyhc_bcQ?si=QgyFy5u3XyfUAPEZ&t=1340). 
They do also include the temperature when they [sample responses](https://github.com/McGill-NLP/nano-aha-moment/blob/0ce62ece2681eefad8041b9336fc7d2cd1a3687a/nano_r1_script.py#L580C4-L592C10). 
I think we could try and add this step here as well. We can see if it's able to improve things.

Comment (Cormac-C) [reply grpo.py:304]: Thanks https://github.com/Cormac-C/grpo-project/pull/11/commits/870f9a2194ecb46d306b4314be97bd27613a3463

**Review by abhijeetpraveen — APPROVED** [review]
Looks good to me!

**Review by Kuzcop — APPROVED** [review]
Looks good!
